### ISSUE TITLE
borderRadius is missing in TS

### ISSUE THREAD

**Issue by appsforartists** [smyte/jsxstyle#92]
I've noticed that `borderRadius` doesn't autocomplete in TS - only `borderTopLeftRadius, borderTopRightRadius, borderBottomLeftRadius, borderBottomRightRadius` do.  `borderRadius` is defined upstream in [`lib.dom.d.ts`'s `CSSStyleDeclaration`](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts).

It looks like you're autogenerating `CSSProperties.ts`, so I'm not sure where/how to fix this.

**Comment by meyer** [comment]
@appsforartists `CSSProperties.ts` is, at the moment, a direct copy of the CSSProperties interface from `@types/react`. The file I’d modify to add `borderRadius` is the [script that generates `CSSProperties.ts`][misc]. The idea of the script is to use React’s CSSProperties as a base and to add on any additional items. I haven’t built out the “add on additional items” portion but it should be pretty straightforward. I’ll try to get to it sometime this weekend.

What I’m probably going to do is to keep a separate `.ts` file in the `misc` folder that contains any additions to `CSSProperties`. When `mktypes.js` runs, it would read that file and merge it in with the types from `@types/react`.

[misc]: https://github.com/smyte/jsxstyle/blob/34de354/misc/mktypes.js

**Comment by appsforartists** [comment]
Thanks for the background!

Seems like a better solution in this case would be to patch [the upstream copy](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts).  Do you expect to need properties in `jsxstyle` that wouldn't make sense in `@types/react`?

**Comment by meyer** [comment]
@appsforartists `borderRadius` definitely should be in the upstream version. As far as jsxstyle goes, in the future their might be a few special case props (for example, `px` to set `paddingLeft` and `paddingRight` to the same value). Still thinking it over though (see #74). We might also feel confident enough in the completeness of jsxstyle’s `CSSProperties` that we’d remove the index accessor (😱).

**Comment by appsforartists** [comment]
Index accessor?

**Comment by meyer** [comment]
this lil fella: [types/react/index.d.ts#L2436](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b485550/types/react/index.d.ts#L2436). index signature maybe? I’m not 100% sure what that thing is called.

---

In addition to my previous comment: I feel like we‘ve got more motive to keep a list of style properties and types accurate and up to date, too.

**Comment by appsforartists** [comment]
Ahhh - I didn't realize `jsxstyle` had one of those.

I left my computer at work, but I'll try to remember to add `borderRadius` upstream, as well as [fix the types](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20743) of all the radii properties sometime soon.

**Comment by meyer** [comment]
Rad, let me know when it lands!

I just realised the other reason I’m using a custom version of `CSSProperties`: I don’t want people to have to install `@types/preact` if they’re using the Preact version of jsxstyle.

**Comment by appsforartists** [comment]
Would it matter?  Types don't ship to users, and you could have `@types/react` as a `devDependency`, so the install would be automated.

Not sure if React and Preact types would ever diverge (and I like the thinking behind shorthand properties), but I don't think depending on `@types/react` would alone be a blocker.

**Comment by appsforartists** [comment]
@meyer https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21895

**Comment by meyer** [comment]
@appsforartists 🙏🙏🙏

**Comment by appsforartists** [comment]
Another thing to consider re: extending vs. forking `@types/react`: if you imported/extended it, you wouldn't have to push new `jsxstyle` releases just to get fixes like this one.

**Comment by meyer** [comment]
@appsforartists yeah, I’m planning to go through the CSS spec and add each value with proper types one of these days when I have a few hours to kill. The CSSProperties interface in `@types/react` seems to be intentionally _not_ comprehensive [according to this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b485550/types/react/index.d.ts#L816-L817):

```ts
// This interface is not complete. Only properties accepting
// unitless numbers are listed here (see CSSProperty.js in React)
```

Seems like their goals don’t 100% align with jsxstyle’s, which is why I looked into the external CSSProperties file in the first place.

Regarding extending vs. copying: if it wasn’t for jsxstyle’s pseudoclass prefixes, I’d be totally down with extending the interface instead of copying it, but since we’d need to generate types for the prefixed style props as well, I’d rather just keep everything on the jsxstyle end. If a comment or type changes on a unprefixed prop, I’d like it to change on the pseudo-prefixed versions as well. Less chance of a weird mismatch.

Also also: `@types/react`’s CSSProperties seems to track a more stable version of the spec, which means that TypeScript marks newer values with partial browser support as invalid. Example: [`alignItems="baseline"`](https://developer.mozilla.org/en-US/docs/Web/CSS/align-items).

**Comment by meyer** [comment]
Over the weekend I looked into converting [MDN’s CSS data](https://github.com/mdn/data) to TypeScript type definitions.

### Pros
1. The data is quite comprehensive and fairly up to date.
2. Someone else is maintaining a database of every CSS property.
3. The data provides a ton of metadata about each property. Generating nicely worded and informative docstring comments should be pretty straightforward.

### Cons
1. The data is not 100% up to date.
2. There doesn’t seem to be a regular publishing schedule for the `mdn-data` NPM module (see https://github.com/mdn/data/issues/132). On our end, we could use a git submodule (😱) or some kind of `curl` operation (😱😱) but it’d be best if we could just use a regularly-published NPM module.
2. Generated types will contain a lot of `any`s. Most CSS properties are too complex to describe with TypeScript types. Best I can do in those cases is to add the `syntax` value in the docstring comment.

**Comment by appsforartists** [comment]
If you use the same API as React's `style` prop, most of your loose types will be `number | string`.  There are cases where `Array` would be nice, but I haven't played with the `jsxstyle` implementation enough to know where they are supported.  (I'm thinking about things like `borderRadius = { [ 8, 0, 0, 8 ] }` which would translate to `8px 0px 0px 8px`).  You'll have a hard time validating string contents with TS, but I think you can be a bit more specific than `any` for most types.

I love the idea of leveraging canonical type definitions.

**Comment by meyer** [comment]
Here’s an example generated type def:

```ts
interface MDNCSSProperties {
  //...

  /**
   * Syntax: visible | hidden
   * Status: standard
   */
  backfaceVisibility?: "visible" | "hidden";

  /**
   * Syntax: [ <bg-layer> , ]* <final-bg-layer>
   * Status: standard
   * Shorthand for background-image, background-position, background-size, background-repeat, background-origin, background-clip, background-attachment, and background-color
   */
  background?: any;

  //...
}
```

Shorthand props are the problematic ones. For example, here’s the syntax definition for `font`:
```
[ [ <'font-style'> || <font-variant-css21> || <'font-weight'> || <'font-stretch'> ]? <'font-size'> [ / <'line-height'> ]? <'font-family'> ] | caption | icon | menu | message-box | small-caption | status-bar
```

Yeaaaaaaaaah that’s gonna be an `any` 😅


I dig array syntax for `borderRadius`. The main reason I haven’t pursued it is because, at the moment, style values of type `object` are unused (they just get stringified). There’s a whole dimension of possibility with object-type style values, but once we add meaning to an object or array value, we’ve locked down that possibility. I want to make sure we lock it down to the right thing.

**Comment by meyer** [comment]
Here’s the rough WIP: https://github.com/smyte/jsxstyle/commit/12f5b3e

**Comment by meyer** [comment]
Here’s the rough WIP commit: https://github.com/smyte/jsxstyle/commit/12f5b3e

and the generated type definitions: [`12f5b3e/packages/jsxstyle/MDNCSSProperties.d.ts`](https://github.com/smyte/jsxstyle/blob/12f5b3e/packages/jsxstyle/MDNCSSProperties.d.ts)

**Comment by appsforartists** [comment]
Since this is CSS, I think you can safely fall back to `string` rather than `any`.  Currently, `array`, `object`, and `boolean` are always wrong.  `number` is appreciated for pixel values, but can always be represented as `16px` and corrected with a future PR.

Personally, I think `number` should _always_ be treated as a pixel value, which means `lineHeight` should accept either a number (in pixels) or a string (in line multiples).  React's inline style prop does the opposite, which leaks the abstraction and makes the style system harder to use.

(Sorry for the side tangent.)

**Comment by meyer** [comment]
@appsforartists yeah, number always being px would be nice (in theory, not sure in practice). A safe option would be to treat `lineHeight` like a unitless prop. Still need a number with props like `flexGrow`.

**Comment by appsforartists** [comment]
Would you be interested in spinning your `mdn-data -> d.ts` project into its own repo?  @kof is down to collaborate on https://github.com/cssinjs/style-type-generator.

**Comment by appsforartists** [comment]
https://www.npmjs.com/package/@types/react has been updated!

**Comment by meyer** [comment]
@appsforartists I just published `jsxstyle@2.0.8` which includes this update. Thanks!
